### PR TITLE
Rest develop

### PR DESCRIPTION
**Pull request by SergeyStroganoff**
This is pull request for discuss a rest service. There are a few problms: HTTP Server  answers with 415 Error for POST and DELETE methods; JWTTokenNeeded does't work. (even I mapped it web.xml). I have no idea about it.

### CODE REVIEW

File: restclient/src/main/java/org/stroganov/restservice/AuthenticationServiceClient.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        JWTToken = response.getEntity(String.class);
+        return JWTToken;
+    }
+
+    public static String getJWTToken() {
+        if (JWTToken != null) {
+            return JWTToken;
+        } else {
+            throw new NullArgumentException("JWTToken empty!");
+        }
```
Comment (Mazel-Tovr) [inline restclient/src/main/java/org/stroganov/restservice/AuthenticationServiceClient.java:42]:
```suggestion
    
            return JWTToken != null ? JWTToken :  throw new NullArgumentException("JWTToken empty!");

```

File: restclient/src/main/java/org/stroganov/restservice/LibraryRestServiceClient.java [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+public class LibraryRestServiceClient implements LibraryDAO {
+    public static final String FAILED_HTTP_ERROR_CODE = "Failed : HTTP error code : ";
+    public static final String USER_PATH = "/user/";
+    private static volatile LibraryRestServiceClient instance;
+    private final Logger logger = Logger.getLogger(LibraryRestServiceClient.class);
+    private final Client client = Client.create();
+    private final Gson gson = new Gson();
+    private final String restServiceURL = PropertiesManager.getProperties().getProperty("restServiceURL");
+
+    public static synchronized LibraryRestServiceClient getInstance() {
```
Comment (Mazel-Tovr) [inline restclient/src/main/java/org/stroganov/restservice/LibraryRestServiceClient.java:27]:
```suggestion
    public static LibraryRestServiceClient getInstance() {
```

File: restserver/src/main/java/org/stroganov/rs/controller/AuthenticationController.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+@Path("/api")
+public class AuthenticationController extends Controller {
+
+    @POST
+    @Path("/authentication")
+    @Consumes({MediaType.APPLICATION_JSON})
+    @Produces({MediaType.APPLICATION_JSON})
+    // QueryParam
+    public Response authentication(User user) {
```
Comment (Mazel-Tovr) [inline restserver/src/main/java/org/stroganov/rs/controller/AuthenticationController.java:28]:
```suggestion
    public Response authentication(UserDto user) {
```

File: restserver/src/main/java/org/stroganov/rs/controller/AuthenticationController.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        if (!ourUser.getPasscodeHash().equals(user.getPasscodeHash())) {
+            return Response.status(401)
+                    .entity("Password is not correct: " + user.getLogin())
+                    .build();
+        }
+        String newJWTToken = createJWTToken(ourUser);
+        final ObjectMapper mapper = new ObjectMapper();
+        ObjectNode json = mapper.createObjectNode();
+        json.put("Bearer", newJWTToken);
+        return Response.status(200)
```
Comment (Mazel-Tovr) [inline restserver/src/main/java/org/stroganov/rs/controller/AuthenticationController.java:40]: Use http status constant from any http library for example HttpStatus.OK
```suggestion
        return Response.status(HttpStatus.OK)
```

Comment (Mazel-Tovr) [reply restserver/src/main/java/org/stroganov/rs/controller/AuthenticationController.java:40]: everywhere

File: restserver/src/main/java/org/stroganov/rs/controller/AuthenticationController.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    private String createJWTToken(User user) {
+        Key key = Keys.secretKeyFor(SignatureAlgorithm.HS256);
+        return Jwts.builder()
+                .setIssuer("LibraryServer")
+                .setIssuedAt(new Date())
+                .setExpiration(Date.from(LocalDateTime.now().plusMinutes(45L).atZone(ZoneId.systemDefault()).toInstant()))
+                .setSubject(user.getLogin())
+                .claim("admin", user.isAdmin())
+                .signWith(key)
```
Comment (Mazel-Tovr) [inline restserver/src/main/java/org/stroganov/rs/controller/AuthenticationController.java:53]:
```suggestion
                .setIssuer(LibraryServer) // constant move to field 
                .setIssuedAt(new Date())
                .setExpiration(Date.from(LocalDateTime.now().plusMinutes(expirationTime).atZone(ZoneId.systemDefault()).toInstant())) // constant move to field 
                .setSubject(user.getLogin())
                .claim(Role.ADMIN.toString(), user.isAdmin())
                .signWith(key)
```

File: restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+@Path("/api/author")
+public class AuthorController extends Controller {
+
+    @POST
+    @Path("/")
+    @Consumes({MediaType.APPLICATION_JSON})
+    @Produces({MediaType.APPLICATION_JSON})
+    @JWTTokenNeeded
+    public Response addAuthor(Author author) throws IOException {
```
Comment (Mazel-Tovr) [inline restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java:19]:
```suggestion
    public Response addAuthor(AuthorDto author) throws IOException {
```

File: restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        boolean operationResult = libraryDAO.addAuthor(author);
+        logger.info("Result of addAuthor: " + operationResult);
+        return Response.status(200)
+                .entity("true")
```
Comment (Mazel-Tovr) [inline restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java:23]:
```suggestion
                .entity("true") ???
```

File: restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                .build();
+    }
+
+    @GET
+    @Path("/{authorName}")
+    @Consumes({MediaType.APPLICATION_JSON})
+    @Produces({MediaType.APPLICATION_JSON})
+    @JWTTokenNeeded
+    public Response findAuthor(@PathParam("authorName") String authorName) {
+        Author author = libraryDAO.findAuthor(authorName);
```
Comment (Mazel-Tovr) [inline restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java:33]:
```suggestion
        AuthorDto author = libraryDAO.findAuthor(authorName);
```

File: restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        return Response.status(200)
+                .entity(author)
```
Comment (Mazel-Tovr) [inline restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java:35]: Receive and response with dto only. 
everywhere

File: restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        Author author = libraryDAO.findAuthor(authorName);
+        if (author == null) {
+            return Response.status(404)
+                    .entity(authorName + " not found")
+                    .build();
+        }
+        boolean operationResult = libraryDAO.deleteAuthorWithAllHisBooks(author);
+        logger.info("Result of deleteAuthorWithAllHisBooks: " + operationResult);
+        return Response.status(200)
+                .entity(operationResult)
```
Comment (Mazel-Tovr) [inline restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java:54]:
```suggestion
                .entity("Delete "+ operationResult ? " successfully" : " not successfully") // meaningful responce
```

File: restserver/src/main/java/org/stroganov/rs/controller/BookController.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+    public Response deleteBook(@PathParam("bookISBN") String bookISBN) {
+        Book book = libraryDAO.findBook(bookISBN);
+        boolean operationResult = false;
+        if (book != null) {
+            try {
+                operationResult = libraryDAO.deleteBook(book);
+            } catch (IOException e) {
+                return Response.status(500)
+                        .entity(e.getMessage())
```
Comment (Mazel-Tovr) [inline restserver/src/main/java/org/stroganov/rs/controller/BookController.java:31]:
```suggestion
                        .entity("server error or smth like this") // never send backend exception message on response ( potentional security vulnerability) 
```

File: restserver/src/main/java/org/stroganov/rs/controller/BookMarkController.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+@Path("/api/bookMark")
+public class BookMarkController extends Controller {
+
+    @DELETE
+    @Path("/")
+    @Consumes({MediaType.APPLICATION_JSON})
+    @Produces({MediaType.APPLICATION_JSON})
+    @JWTTokenNeeded
+    public Response deleteBookMark(BookMark bookMark) {
```
Comment (Mazel-Tovr) [inline restserver/src/main/java/org/stroganov/rs/controller/BookMarkController.java:22]:
```suggestion
    public Response deleteBookMark(@PathParam ....) {
```

File: restserver/src/main/java/org/stroganov/rs/controller/BookMarkController.java [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+
+@Path("/api/bookMark")
+public class BookMarkController extends Controller {
+
+    @DELETE
+    @Path("/")
```
Comment (Mazel-Tovr) [inline restserver/src/main/java/org/stroganov/rs/controller/BookMarkController.java:18]:
```suggestion
    @Path("/") //Path param
```

File: restserver/src/main/java/org/stroganov/rs/controller/BookMarkController.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+                .entity(operationResult)
+                .build();
+    }
+
+    @POST
+    @Path("/")
+    @Consumes({MediaType.APPLICATION_JSON})
+    @Produces({MediaType.APPLICATION_JSON})
+    @JWTTokenNeeded
+    public Response addBookMark(BookMark bookMark) {
```
Comment (Mazel-Tovr) [inline restserver/src/main/java/org/stroganov/rs/controller/BookMarkController.java:41]:
```suggestion
    public Response addBookMark(BookMarkDto bookMark) {
```

**Review by Mazel-Tovr — CHANGES_REQUESTED** [review]

**Comment by Mazel-Tovr** [comment]
and remove code duplication on client

File: restclient/src/main/java/org/stroganov/restservice/AuthenticationServiceClient.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+import org.stroganov.util.PropertiesManager;
+
+public class AuthenticationServiceClient {
+
+    public static final String FAILED_HTTP_ERROR_CODE = "Failed : HTTP error code : ";
+    private final Logger logger = Logger.getLogger(AuthenticationServiceClient.class);
+    private static final Client client = Client.create();
+    private static final Gson gson = new Gson();
+    private static final String REST_SERVICE_AUTHENTICATION_URL = PropertiesManager.getProperties().getProperty("restServiceAuthenticationURL");
+    private static String JWTToken;
```
[the remark this answers is not in the document]
Comment (SergeyStroganoff) [reply restclient/src/main/java/org/stroganov/restservice/AuthenticationServiceClient.java:19]: Fixed.

File: restclient/src/main/java/org/stroganov/restservice/AuthenticationServiceClient.java (hunk 1 above)
Comment (SergeyStroganoff) [reply restclient/src/main/java/org/stroganov/restservice/AuthenticationServiceClient.java:42]: unfortunately it does't work with : throw new ...

File: restclient/src/main/java/org/stroganov/restservice/LibraryRestServiceClient.java (hunk 2 above)
Comment (SergeyStroganoff) [reply restclient/src/main/java/org/stroganov/restservice/LibraryRestServiceClient.java:27]: fixed.

File: restserver/src/main/java/org/stroganov/rs/controller/AuthenticationController.java (hunk 3 above)
Comment (SergeyStroganoff) [reply restserver/src/main/java/org/stroganov/rs/controller/AuthenticationController.java:28]: In this method i have code to check for compliance gotten password with pasword of USer in DB:
 User ourUser = libraryDAO.findUser(user.getLogin());
        if (!ourUser.getPasscodeHash().equals(user.getPasscodeHash())) {
            return Response.status(401)
                    .entity("Password is not correct: " + user.getLogin())
                    .build();
        }
But UserDTO does't have a psssword - how will we check password?

File: restserver/src/main/java/org/stroganov/rs/controller/AuthenticationController.java (hunk 4 above)
Comment (SergeyStroganoff) [reply restserver/src/main/java/org/stroganov/rs/controller/AuthenticationController.java:40]: Fixed.

File: restserver/src/main/java/org/stroganov/rs/controller/AuthenticationController.java (hunk 5 above)
Comment (SergeyStroganoff) [reply restserver/src/main/java/org/stroganov/rs/controller/AuthenticationController.java:53]: Fixed.

File: restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java (hunk 7 above)
Comment (SergeyStroganoff) [reply restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java:23]: fixed.

File: restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java (hunk 10 above)
Comment (SergeyStroganoff) [reply restserver/src/main/java/org/stroganov/rs/controller/AuthorController.java:54]: But in client i suggest true ore false ...

File: restserver/src/main/java/org/stroganov/rs/controller/BookController.java (hunk 11 above)
Comment (SergeyStroganoff) [reply restserver/src/main/java/org/stroganov/rs/controller/BookController.java:31]: fixed

File: restserver/src/main/java/org/stroganov/rs/controller/BookMarkController.java (hunk 12 above)
Comment (SergeyStroganoff) [reply restserver/src/main/java/org/stroganov/rs/controller/BookMarkController.java:22]: fixed.

File: restserver/src/main/java/org/stroganov/rs/controller/BookMarkController.java (hunk 13 above)
Comment (SergeyStroganoff) [reply restserver/src/main/java/org/stroganov/rs/controller/BookMarkController.java:18]: fixed

File: restclient/src/main/java/org/stroganov/dao/DAOFactory.java [hunk 16]
```diff
@@ -0,0 +1,15 @@
+package org.stroganov.dao;
+
+import org.stroganov.exceptions.DBExceptions;
+
+public class DAOFactory {
```
Comment (Mazel-Tovr) [inline restclient/src/main/java/org/stroganov/dao/DAOFactory.java:5]: This class looks like useless

File: restclient/src/main/java/org/stroganov/dialogue/MenuManagerDialogue.java [hunk 17]
```diff
@@ -330,19 +331,25 @@ public boolean deleteAuthor() {
...
                 userInterface.showMessage("An error happen:" + e.getMessage());
                 return false;
             }
-            String successMessage = "Author " + authorName + "was successfully deleted with all his books";
+            String successMessage;
+            if (operationResult) {
+                successMessage = AUTHOR + authorName + "was successfully deleted with all his books";
+            } else {
+                successMessage = AUTHOR + authorName + "was not successfully deleted. See logs";
+            }
```
Comment (Mazel-Tovr) [inline restclient/src/main/java/org/stroganov/dialogue/MenuManagerDialogue.java:347]:
```suggestion
            String successMessage = AUTHOR + authorName + (operationResult ? "was successfully deleted with all his books" : "was not successfully deleted. See logs" );
```

File: restserver/src/main/java/org/stroganov/dao/JsonLibraryDAO.java [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+    private final List<BookMark> bookMarkList;
+    private final List<Author> authorsList;
+    private final JsonDBSaver jsonDBSaver;
+
+    public static synchronized JsonLibraryDAO getInstance() throws DBExceptions {
+        if (instance == null) {
+            instance = new JsonLibraryDAO();
+        }
+        return instance;
+    }
```
Comment (Mazel-Tovr) [inline restserver/src/main/java/org/stroganov/dao/JsonLibraryDAO.java:31]: Use double check singleton pattern

**Review by Mazel-Tovr — COMMENTED** [review]
Please, when you fix comment press on
"Resolve conversation" button, no need to write fixed

**Review by Mazel-Tovr — CHANGES_REQUESTED** [review]
1. Code formatting in (AuthenticationServiceClient.java)
2. As I write before, response entity should be dto everywhere. You receive dto, and response with dto/simple text/no body etc but not with object that uses only in backend logic
Fix this and approve

**Comment by SergeyStroganoff** [comment]
Merging after cheking and refactoring.

File: apidtoobjects/src/main/java/org/stroganov/util/TransitionObjectsService.java [hunk 19]
```diff
@@ -49,4 +52,22 @@ public static History getHistory(HistoryDTO historyDTO) {
         history.setUser(getUser(historyDTO.getUser()));
         return history;
     }
+
+    public static List<BookMarkDTO> getBookMarkDTOList(List<BookMark> bookMarkList) {
+        List<BookMarkDTO> bookMarkDTOList = new ArrayList<>();
+        bookMarkList.forEach(bookMark -> bookMarkDTOList.add(new BookMarkDTO(bookMark)));
+        return bookMarkDTOList;
```
Comment (Mazel-Tovr) [inline apidtoobjects/src/main/java/org/stroganov/util/TransitionObjectsService.java:59]: Use `map` function from stream api
`return bookMarkList.stream().map(e -> new BookMarkDTO(у)).collect(Collectors.toList());`

**Pull request merged**